MQTT 3.1.1 and 5.0 transport binding

clemensv opened the pull request:
Proposed MQTT 3.1.1 and 5.0 transport binding, analogous to the same for [HTTP](https://github.com/cloudevents/spec/pull/148) and AMQP. This includes the same JSON mapping as the HTTP PR for completeness and the JSON mapping should be discussed there, since HTTP has higher priority.

Signed-off-by: Clemens Vasters <clemensv@microsoft.com>

lfourie commented:
Change  http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1html
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html

lfourie commented:
Change http://docs.oasis-open.org/mqtt/mqtt/5.0.0/mqtt-5.0.0.html
to http://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+For instance, if the declared `contentType` is
+`application/json;charset=utf-8`, the expectation is that the `data` attribute
+value is made available as [UTF-8][RFC3629] encoded JSON text for use in
+AMQP.
```
Change AMQP to MQTT

clemensv replied:
Done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+## 3. MQTT PUBLISH Message Mapping
+
+With MQTT 5.0, the content mode is chosen by the sender of the event. Gestures
+that might allow solicitation of events using a particular mode might be
```
"Gestures .... " is unclear - reword sentence.

clemensv replied:
"Protocol usage patterns"

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+that might allow solicitation of events using a particular mode might be
+defined by an application, but are not defined here.
+
+The receiver of the event can distinguish between the two modes by inspecting
```
change to "two content modes"

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+defined by an application, but are not defined here.
+
+The receiver of the event can distinguish between the two modes by inspecting
+the `content-type` message field. If the value is prefixed with the Cloud
```
unclear what "value" is referring to. can you clarify.

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+[event format](#14-event-formats), the receiver uses *structured* mode,
+otherwise it defaults to *binary* mode.
+
+If a receiver detects the Cloud Events media type, but with an event format that
```
Can you clarify how the receiver "detects the Cloud Event media type".

clemensv replied:
"If a receiver finds a CloudEvents media type as per the above rule"

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+defined by an application, but are not defined here.
+
+The receiver of the event can distinguish between the two modes by inspecting
+the `content-type` message field. If the value is prefixed with the Cloud
```
Replace " the 'content-type' message field" with "the Content Type of the MQTT PUBLISH message"

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+
+#### 3.1.1 MQTT PUBLISH content-type
+
+For the *binary* mode, the MQTT PUBLISH message's [`content-type`][5-content-type]
```
Replace "content-type" with "MQTT Content Type property"

clemensv replied:
"the MQTT PUBLISH message's [`Content Type`][5-content-type] property"

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+hops, and across multiple transports. This is the only supported mode for 
+MQTT 3.1.1
+
+#### 3.2.1. MQTT Content-Type
```
"Content Type" not "Content-Type"

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+
+#### 3.2.1. MQTT Content-Type
+
+For MQTT 5.0, the [MQTT `content-type`][5-content-type] field is set to the media
```
Content Type property

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+#### 3.1.3. Metadata Headers
+
+All [Cloud Events][CE] attributes with exception of `contentType` and `data`
+are individually mapped to and from the MQTT PUBLISH User Property section.
```
change to: ... from one or more User Property fields in the MQTT PUBLISH message.

clemensv replied:
I am making this slightly stringer: "MUST be individually mapped to and from the User Property fields in the MQTT PUBLISH message."

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+
+##### 3.1.3.1 User Property Names
+
+Cloud Event attribute names are used unchanged in the MQTT PUBLISH User Property
```
change to "in each User Property in the MQTT PUBLISH message.

clemensv replied:
Again, stronger: "Cloud Event attribute names MUST be used unchanged in each mapped User Property
in the MQTT PUBLISH message."

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+
+##### 3.1.3.2 User Property Values
+
+The value for each MQTT PUBLISH user property is constructed from the respective
```
User Property

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
````diff
@@ -0,0 +1,272 @@
+``` text
+------------------ PUBLISH -------------------
+
+topic: mytopic
````
topic: -> Topic Name:

clemensv replied:
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+[CE]: ./spec.md
+[JSON-format]: ./json-format.md
+[OASIS-MQTT-3.1.1]: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1html
+[OASIS-MQTT-5]: http://docs.oasis-open.org/mqtt/mqtt/5.0.0/mqtt-5.0.0.html
```
http://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html

clemensv replied:
fixed

clemensv commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+The *binary* content mode accommodates any shape of event data, and allows for
+efficient transfer and without transcoding effort.
+
+#### 3.1.1 MQTT PUBLISH content-type
```
done

clemensv commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+------------------ PUBLISH -------------------
+
+topic: mytopic
+content-type: application/json; charset=utf-8
```
done

clemensv commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+
+#### 3.2.3. Metadata Headers
+
+For MQTT 5.0, implementations MAY include the same MQTT PUBLISH user properties
```
done

clemensv commented on mqtt-transport-binding.md:
````diff
@@ -0,0 +1,272 @@
+``` text
+------------------ PUBLISH -------------------
+
+topic: mytopic
````
done

clemensv commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,272 @@
+------------------ PUBLISH -------------------
+
+topic: mytopic
+content-type: application/json; charset=utf-8
```
done

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,277 @@
+[RFC2046]: https://tools.ietf.org/html/rfc2046
+[RFC2119]: https://tools.ietf.org/html/rfc2119
+[RFC3629]: https://tools.ietf.org/html/rfc3629
+[RFC4627]: https://tools.ietf.org/html/rfc4627
```
Add: [RFC7159] https://tools.ietf.org/html/rfc7159

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,277 @@
+[CloudEvents][CE] is a standardized and transport-neutral definition of the
+structure and metadata description of events. This specification defines how
+the elements defined in the CloudEvents specification are to be used in
+MQTT PUBLISH ([3.1.1][3-publish], [5.0][5-publish]) messages.
```
Fix links

lfourie commented on mqtt-transport-binding.md:
```diff
@@ -0,0 +1,277 @@
+
+This specification does not prescribe rules constraining transfer or
+settlement of event messages with MQTT; it solely defines how CloudEvents
+are expressed as MQTT PUBLISH messages ([3.1.1][3-publish], [5.0][5-publish]).
```
Fix links

jwende commented:
Not sure if this is by intention but within the documents there are different notations for content type:
application/json;charset=utf-8
application/jcloudevents+json;charset=utf-8
application/cloudevents+json; charset=UTF-8

jwende commented:
3.1.4 Examples - I would suggest to use a mapping table between MQTT V5 PUBLISH message fields and the content of the cloud event

jwende commented:
3.2.4 Examples  - I would prefer two examples - one for MQTT 3.1.1 and one for MQTT 5.0

jwende commented:
3.2.3. Metadata Headers - I would vote for SHOULD instead of MAY - to keep it as close as possible to the binary content mode

clemensv commented:
@jwende I've addressed your comments in an update, mostly with clarifications. Exception: I left "MAY" stand for the moment, mostly because I've got "MAY" in all bindings and if we change that I want to make it one change and I want the group to agree to us duplicating the metadata.

jwende commented:
@clemensv thanks. Will take a look later today.

yaronha commented:
@clemensv following the call, would be good to have the same attribute names (in binary mode) for all transport, seems like HTTP and MQTT attr names dont match, i would like to avoid attribute+transport specific consumer logic

duglin commented:
@clemensv I rebased (and fixed the conflict in Makefile) for you - hope that's ok.
